src/audio/rsp_mixer.S

```diff
@@ -0,0 +1,1133 @@
+	lqv v_chvol_l_2,0,     0*MAX_CHANNELS_VOFF+2,s0
+	lqv v_chvol_l_3,0,     0*MAX_CHANNELS_VOFF+3,s0
+
+	lqv v_chvol_r_0,0,     1*MAX_CHANNELS_VOFF+0,s0
+	lqv v_chvol_r_1,0,     1*MAX_CHANNELS_VOFF+1,s0
+	lqv v_chvol_r_2,0,     1*MAX_CHANNELS_VOFF+2,s0
+	lqv v_chvol_r_3,0,     1*MAX_CHANNELS_VOFF+3,s0
+
+	# Apply global volume to obtain the final volume for each channel
+	vmudl v_chvol_l_0, v_chvol_l_0, v_glvol,0
+	vmudl v_chvol_r_0, v_chvol_r_0, v_glvol,0
+	vmudl v_chvol_l_1, v_chvol_l_1, v_glvol,0
+	vmudl v_chvol_r_1, v_chvol_r_1, v_glvol,0
+	vmudl v_chvol_l_2, v_chvol_l_2, v_glvol,0
+	vmudl v_chvol_r_2, v_chvol_r_2, v_glvol,0
+	vmudl v_chvol_l_3, v_chvol_l_3, v_glvol,0
+	vmudl v_chvol_r_3, v_chvol_r_3, v_glvol,0
+
+#if VOLUME_FILTER
+	# Load actual volumes levels
+	lqv v_xvol_l_0,0,      0*MAX_CHANNELS_VOFF+0,s1
+	lqv v_xvol_l_1,0,      0*MAX_CHANNELS_VOFF+1,s1
+	lqv v_xvol_l_2,0,      0*MAX_CHANNELS_VOFF+2,s1
+	lqv v_xvol_l_3,0,      0*MAX_CHANNELS_VOFF+3,s1
+
+	lqv v_xvol_r_0,0,      1*MAX_CHANNELS_VOFF+0,s1
+	lqv v_xvol_r_1,0,      1*MAX_CHANNELS_VOFF+1,s1
+	lqv v_xvol_r_2,0,      1*MAX_CHANNELS_VOFF+2,s1
+	lqv v_xvol_r_3,0,      1*MAX_CHANNELS_VOFF+3,s1
+#else
+	vor v_xvol_l_0, v_chvol_l_0, v_zero,0
+	vor v_xvol_l_1, v_chvol_l_1, v_zero,0
+	vor v_xvol_l_2, v_chvol_l_2, v_zero,0
+	vor v_xvol_l_3, v_chvol_l_3, v_zero,0
+
+	vor v_xvol_r_0, v_chvol_r_0, v_zero,0
+	vor v_xvol_r_1, v_chvol_r_1, v_zero,0
+	vor v_xvol_r_2, v_chvol_r_2, v_zero,0
+	vor v_xvol_r_3, v_chvol_r_3, v_zero,0
+#endif
+
+	jr ra
+	nop
+	.endfunc
+
+	#undef v_glvol 
+
+
+	.func EndMixer
+EndMixer:
+	li s1, %lo(XVOL_L)
+
+	sqv v_xvol_l_0,0,      0*MAX_CHANNELS_VOFF+0,s1
+	sqv v_xvol_l_1,0,      0*MAX_CHANNELS_VOFF+1,s1
+	sqv v_xvol_l_2,0,      0*MAX_CHANNELS_VOFF+2,s1
+	sqv v_xvol_l_3,0,      0*MAX_CHANNELS_VOFF+3,s1
+
+	sqv v_xvol_r_0,0,      1*MAX_CHANNELS_VOFF+0,s1
+	sqv v_xvol_r_1,0,      1*MAX_CHANNELS_VOFF+1,s1
+	sqv v_xvol_r_2,0,      1*MAX_CHANNELS_VOFF+2,s1
+	sqv v_xvol_r_3,0,      1*MAX_CHANNELS_VOFF+3,s1
+
+	jr ra
+	nop
+
+	.endfunc
+
+
+##############################################################
+# Mixer
+#
+# Arguments:
+#    s4:  buffer into which the mixed samples will be stored
+#    k1:  num samples
+#
+##############################################################
+
+	#define v_out_l       $v01
+	#define v_out_r       $v02
+	#define v_sample_0    $v03
+	#define v_sample_1    $v04
+	#define v_sample_2    $v05
+	#define v_sample_3    $v06
+	#define v_mix_l       $v07
+	#define v_mix_r       $v08
+
+	.func Mixer
+Mixer:
+	# Load samples
+	li s0, %lo(CHANNEL_BUFFER)
+	move t1, k1
+
+	# Load initial samples
+	lqv v_sample_0,0, 0,s0
+	lqv v_sample_1,0, 1,s0
+	lqv v_sample_2,0, 2,s0
+	lqv v_sample_3,0, 3,s0
+
+	# For optimal pipelining, output is stored at the beginning of the loop. To avoid
+	# corrupting memory, load the output register with whatever is there now.
+	lsv v_out_l,0, -2,s4
+	blt k0, 8, Mix8Start    # Optimized mixing loop for <= 8 channels
```

Comment:
The comment says `<= 8` but this actually checks for `< 8`

Reply:
Good catch, I'll fix this